File: Swift2/Sources/Swift2/HexagramValueSequencer.swift

public class CHexagramValueSequencer : CValueSequencer {
	
	init(nValue: Int = 0) {
		super.init(nInnerSequencers: 2, nValues: 64, nValue: nValue)
		m_pvsInner = [CTrigramValueSequencer(nValue: 0) as CValueSequencer,CTrigramValueSequencer(nValue: 0) as CValueSequencer]
		//m_pvsInner[1] = new CTrigramValueSequencer(-1)
		//m_pvsInner[0] = new CTrigramValueSequencer(-1)
		Trigram(nIndex: 1).SetParent(pvsParent: self)
		Trigram(nIndex: 0).SetParent(pvsParent: self)
		m_nSequences = Sequences.nHexagramSequences
		m_nRatios = Sequences.nHexagramRatios
		Value = nValue
		UpdateInnerValues()
		//UpdateOuterValues()
	}

	convenience init(strValue: String) {
		let s = strValue.trimmingCharacters(in: [" "])
		if let fValue = Double(s) {
			self.init(nValue: Int(fValue))
			let s1 = s.split(separator: ".")
			if s1.count == 2 {
				var line = -1
				for l in "123456" {
					line += 1
					if s1[1].contains(l) {
						Trigram(nIndex: line / 3).Line(nIndex: line % 3).Next()
					}
				}
			}
		}
		else {
			self.init()
		}
	}

	convenience init(hvs: CHexagramValueSequencer) {
		self.init() // (nInnerSequencers: 2, nValues: 64, nValue: 0)
		Trigram(nIndex: 1).Line(nIndex: 2).Value = hvs.Trigram(nIndex: 1).Line(nIndex: 2).Value
		Trigram(nIndex: 1).Line(nIndex: 1).Value = hvs.Trigram(nIndex: 1).Line(nIndex: 1).Value
		Trigram(nIndex: 1).Line(nIndex: 0).Value = hvs.Trigram(nIndex: 1).Line(nIndex: 0).Value
		Trigram(nIndex: 0).Line(nIndex: 2).Value = hvs.Trigram(nIndex: 0).Line(nIndex: 2).Value
		Trigram(nIndex: 0).Line(nIndex: 1).Value = hvs.Trigram(nIndex: 0).Line(nIndex: 1).Value
		Trigram(nIndex: 0).Line(nIndex: 0).Value = hvs.Trigram(nIndex: 0).Line(nIndex: 0).Value

		Trigram(nIndex: 1).Line(nIndex: 2).UpdateInnerValues()
		Trigram(nIndex: 1).Line(nIndex: 1).UpdateInnerValues()
		Trigram(nIndex: 1).Line(nIndex: 0).UpdateInnerValues()
		Trigram(nIndex: 0).Line(nIndex: 2).UpdateInnerValues()
		Trigram(nIndex: 0).Line(nIndex: 1).UpdateInnerValues()
		Trigram(nIndex: 0).Line(nIndex: 0).UpdateInnerValues()

		Trigram(nIndex: 1).Line(nIndex: 2).UpdateOuterValues()
		Trigram(nIndex: 1).Line(nIndex: 1).UpdateOuterValues()
		Trigram(nIndex: 1).Line(nIndex: 0).UpdateOuterValues()
		Trigram(nIndex: 0).Line(nIndex: 2).UpdateOuterValues()
		Trigram(nIndex: 0).Line(nIndex: 1).UpdateOuterValues()
		Trigram(nIndex: 0).Line(nIndex: 0).UpdateOuterValues()

		//UpdateInnerValues()
		//UpdateOuterValues()
	}

	public func Trigram(nIndex: Int) -> CTrigramValueSequencer {
 		return m_pvsInner[nIndex] as! CTrigramValueSequencer
	}

	@discardableResult
	public override func Inverse() -> CValueSequencer {
		var bMoving: [[Bool]] = [[ false, false, false ], [ false, false, false ]]
		SaveMoving(bMoving: &bMoving)
		Value = ((Trigram(nIndex: 1).Line(nIndex: 2).Value % 2 == 0 ? 0 : 1) +
			((Trigram(nIndex: 1).Line(nIndex: 1).Value % 2 == 0 ? 0 : 1) * 2) +
			((Trigram(nIndex: 1).Line(nIndex: 0).Value % 2 == 0 ? 0 : 1) * 4)) +
			((((Trigram(nIndex: 0).Line(nIndex: 2).Value % 2 == 0 ? 0 : 1) +
			((Trigram(nIndex: 0).Line(nIndex: 1).Value % 2 == 0 ? 0 : 1) * 2) +
			((Trigram(nIndex: 0).Line(nIndex: 0).Value % 2 == 0 ? 0 : 1) * 4))) * 8)
		UpdateInnerValues()
		UpdateOuterValues()
		RestoreMoving(bMoving: &bMoving,bInverseTrigram: true, bInverseLine: true)
		return self
	}

	@discardableResult
	public override func Opposite() -> CValueSequencer {
		var bMoving: [[Bool]] = [[ false, false, false ], [ false, false, false ]]
		SaveMoving(bMoving: &bMoving)
		super.Opposite()
		RestoreMoving(bMoving: &bMoving, bInverseTrigram: false, bInverseLine: false)
		return self
	}

	@discardableResult
	public override func Transverse() -> CValueSequencer {
		var nTemp: Int
		var bMoving: [[Bool]] = [[ false, false, false ], [ false, false, false ]]
		SaveMoving(bMoving: &bMoving)
		nTemp = Trigram(nIndex: 0).Value
		Trigram(nIndex: 0).Value = Trigram(nIndex: 1).Value
		Trigram(nIndex: 1).Value = nTemp
		Value = Trigram(nIndex: 0).Value + Trigram(nIndex: 1).Value * 8
		UpdateInnerValues()
		UpdateOuterValues()
		RestoreMoving(bMoving: &bMoving, bInverseTrigram: true, bInverseLine: false)
		return self
	}

	@discardableResult
	public override func Nuclear() -> CValueSequencer {
		var nTemp: Int
		var bMoving: [[Bool]] = [[ false, false, false ], [ false, false, false ]]
		SaveMoving(bMoving: &bMoving)
		nTemp = Trigram(nIndex: 1).Line(nIndex: 0).Value
		Trigram(nIndex: 1).Line(nIndex: 2).Value = Trigram(nIndex: 1).Line(nIndex: 1).Value
		Trigram(nIndex: 1).Line(nIndex: 1).Value = Trigram(nIndex: 1).Line(nIndex: 0).Value
		Trigram(nIndex: 1).Line(nIndex: 0).Value = Trigram(nIndex: 0).Line(nIndex: 2).Value
		Trigram(nIndex: 0).Line(nIndex: 0).Value = Trigram(nIndex: 0).Line(nIndex: 1).Value
		Trigram(nIndex: 0).Line(nIndex: 1).Value = Trigram(nIndex: 0).Line(nIndex: 2).Value
		Trigram(nIndex: 0).Line(nIndex: 2).Value = nTemp
		Value = ((((Trigram(nIndex: 1).Line(nIndex: 2).Value % 2 == 0 ? 0 : 1) * 4) +
			((Trigram(nIndex: 1).Line(nIndex: 1).Value % 2 == 0 ? 0 : 1) * 2) +
			(Trigram(nIndex: 1).Line(nIndex: 0).Value % 2 == 0 ? 0 : 1)) * 8) +
			((Trigram(nIndex: 0).Line(nIndex: 2).Value % 2 == 0 ? 0 : 1) * 4) +
			((Trigram(nIndex: 0).Line(nIndex: 1).Value % 2 == 0 ? 0 : 1) * 2) +
			(Trigram(nIndex: 0).Line(nIndex: 0).Value % 2 == 0 ? 0 : 1)
		UpdateInnerValues()
		UpdateOuterValues()
		if bMoving[1][1] {
			Trigram(nIndex: 1).Line(nIndex: 2).Old()
		}
		if bMoving[1][0] {
			Trigram(nIndex: 1).Line(nIndex: 0).Old()
			Trigram(nIndex: 1).Line(nIndex: 1).Old()
		}
		if bMoving[0][2] {
			Trigram(nIndex: 0).Line(nIndex: 2).Old()
			Trigram(nIndex: 0).Line(nIndex: 1).Old()
		}
		if bMoving[0][1] {
			Trigram(nIndex: 0).Line(nIndex: 0).Old()
		}
		return self
	}

	@discardableResult
	public override func Move() -> CValueSequencer {
		Trigram(nIndex: 1).Move()
		Trigram(nIndex: 0).Move()
		return self
	}

	@discardableResult
	public override func Young() -> CValueSequencer {
		Trigram(nIndex: 1).Young()
		Trigram(nIndex: 0).Young()
		return self
	}

	public override func UpdateInnerValues() {
		Trigram(nIndex: 1).Value = Value / 8
		Trigram(nIndex: 0).Value = Value % 8
		Trigram(nIndex: 1).UpdateInnerValues()
		Trigram(nIndex: 0).UpdateInnerValues()
	}

	public override func UpdateOuterValues() {
		Value = Trigram(nIndex: 0).Value + (Trigram(nIndex: 1).Value * 8)
	}

	private func SaveMoving(bMoving: inout [[Bool]]) {
		for t in 0 ... 1 {
			for l in 0 ... 2 {
				if (Trigram(nIndex: t).Line(nIndex: l).Value == 0) || (Trigram(nIndex: t).Line(nIndex: l).Value == 3) {
					bMoving[t][l] = true
				}
			}
		}
	}

	private func RestoreMoving(bMoving: inout [[Bool]], bInverseTrigram: Bool, bInverseLine: Bool) {
		for t in 0 ... 1 {
			for l in 0 ... 2 {
				if bMoving[t][l] {
					var l1 = l
					if bInverseLine {
						if (l1 == 0) {
							l1 = 2
						}
						else
							if l1 == 2 {
								l1 = 0
							}
					}
					Trigram(nIndex: bInverseTrigram ? 1 - t : t).Line(nIndex: l1).Old()
				}
			}
		}
	}

	public static func SetCurrentSequence(nSequence: Int) {
		m_nCurrentSequence = nSequence
	}

	public static func SetCurrentRatio(nRatio: Int) {
		m_nCurrentRatio = nRatio
	}

	public static func SetCurrentLabel(nLabel: Int) {
		m_nCurrentLabel = nLabel
	}

	public override func GetLabel() -> String { 
		return Sequences.strHexagramLabels[GetCurrentLabel()][Value]
	}

	public override func GetMoving() -> Bool {
		return Trigram(nIndex: 0).IsMoving || Trigram(nIndex: 1).IsMoving
	}

	public func HexagramId(bValueId: Bool = false) -> String {
		var s = String(format: "%2d", bValueId ?  Value : Sequence + 1)
		if IsMoving	{
			s = s + "."
			for l in 0 ... 5 {
				if Trigram(nIndex: l / 3).Line(nIndex: l % 3).IsMoving {
					s = s + String(l + 1)
				}
			}
		}
		return s;
	}

	public func DescribePrimary(bValueId: Bool = false, bIncludeValue: Bool = false) -> String {
		return HexagramId(bValueId: bValueId) + " " + Label + (bIncludeValue ? " (" + ValueStr + ")" : "")
	}

	public func DescibeSecondary(bValueId: Bool = false, bIncludeValue: Bool = false) -> String {
		if IsMoving	{
			let hvsPrimary: CHexagramValueSequencer = self
			let hvsSeconday: CHexagramValueSequencer = CHexagramValueSequencer(hvs: hvsPrimary)
			hvsSeconday.Move()
			return hvsSeconday.HexagramId(bValueId: bValueId) + " " + hvsSeconday.Label + (bIncludeValue ? " (" + hvsSeconday.ValueStr + ")" : "")
		}
		return ""
	}

	public func DescribeCast(bValueId: Bool = false, bIncludeValue: Bool = false) -> String	{
		return DescribePrimary(bValueId: bValueId, bIncludeValue: bIncludeValue) + (IsMoving ? " > " +
			DescibeSecondary(bValueId: bValueId, bIncludeValue: bIncludeValue) : "")
	}

	public override func GetCurrentSequence() -> Int { return CHexagramValueSequencer.m_nCurrentSequence }
	public override func GetCurrentRatio() -> Int { return CHexagramValueSequencer.m_nCurrentRatio }
	public override func GetCurrentLabel() -> Int { return CHexagramValueSequencer.m_nCurrentLabel }

	private static var m_nCurrentSequence: Int = 0
	private static var m_nCurrentRatio: Int = 0
	private static var m_nCurrentLabel: Int = 0
}

File: ValueSequencer.py/hexagram_value_sequencer.py
from __future__ import annotations

from typing import List, Optional

from .trigram_value_sequencer import CTrigramValueSequencer
from .value_sequencer import CValueSequencer
from .sequences import Sequences


class CHexagramValueSequencer(CValueSequencer):
    _nCurrentSequence = 0
    _nCurrentRatio = 0
    _nCurrentLabel = 0

    def __init__(
        self,
        value: int = 0,
        copy_from: Optional["CHexagramValueSequencer"] = None,
    ) -> None:
        super().__init__(2, 64, value if copy_from is None else 0)
        self._pvsInner = [
            CTrigramValueSequencer(-1),
            CTrigramValueSequencer(-1),
        ]
        self.trigram(1).set_parent(self)
        self.trigram(0).set_parent(self)
        self._nSequences = Sequences.nHexagramSequences
        self._nRatios = Sequences.nHexagramRatios
        if copy_from is not None:
            self._copy_from(copy_from)
            self.update_outer_values()
        else:
            self.value = value
            self.update_inner_values()

    def _copy_from(self, source: "CHexagramValueSequencer") -> None:
        for t in range(2):
            for l in range(3):
                line = self.trigram(t).line(l)
                source_line = source.trigram(t).line(l)
                line.value = source_line.value
                line.update_inner_values()
                line.update_outer_values()

    def trigram(self, index: int) -> CTrigramValueSequencer:
        return self._pvsInner[index]  # type: ignore[return-value]

    def inverse(self) -> "CHexagramValueSequencer":
        moving = self._save_moving()
        self.value = (
            (self.trigram(1).line(2).value % 2) +
            (self.trigram(1).line(1).value % 2) * 2 +
            (self.trigram(1).line(0).value % 2) * 4 +
            (
                (
                    (self.trigram(0).line(2).value % 2) +
                    (self.trigram(0).line(1).value % 2) * 2 +
                    (self.trigram(0).line(0).value % 2) * 4
                )
                * 8
            )
        )
        self.update_inner_values()
        self.update_outer_values()
        self._restore_moving(moving, True, True)
        return self

    def opposite(self) -> "CHexagramValueSequencer":
        moving = self._save_moving()
        super().opposite()
        self._restore_moving(moving, False, False)
        return self

    def transverse(self) -> "CHexagramValueSequencer":
        moving = self._save_moving()
        temp = self.trigram(0).value
        self.trigram(0).value = self.trigram(1).value
        self.trigram(1).value = temp
        self.value = self.trigram(0).value + self.trigram(1).value * 8
        self.update_inner_values()
        self.update_outer_values()
        self._restore_moving(moving, True, False)
        return self

    def nuclear(self) -> "CHexagramValueSequencer":
        moving = self._save_moving()
        temp = self.trigram(1).line(0).value
        self.trigram(1).line(2).value = self.trigram(1).line(1).value
        self.trigram(1).line(1).value = self.trigram(1).line(0).value
        self.trigram(1).line(0).value = self.trigram(0).line(2).value
        self.trigram(0).line(0).value = self.trigram(0).line(1).value
        self.trigram(0).line(1).value = self.trigram(0).line(2).value
        self.trigram(0).line(2).value = temp
        self.value = (
            (
                (self.trigram(1).line(2).value % 2) * 4 +
                (self.trigram(1).line(1).value % 2) * 2 +
                (self.trigram(1).line(0).value % 2)
            )
            * 8
            + (self.trigram(0).line(2).value % 2) * 4
            + (self.trigram(0).line(1).value % 2) * 2
            + (self.trigram(0).line(0).value % 2)
        )
        self.update_inner_values()
        self.update_outer_values()
        if moving[1][1]:
            self.trigram(1).line(2).old()
        if moving[1][0]:
            self.trigram(1).line(0).old()
            self.trigram(1).line(1).old()
        if moving[0][2]:
            self.trigram(0).line(2).old()
            self.trigram(0).line(1).old()
        if moving[0][1]:
            self.trigram(0).line(0).old()
        return self

    def move(self) -> "CHexagramValueSequencer":
        self.trigram(1).move()
        self.trigram(0).move()
        return self

    def young(self) -> "CHexagramValueSequencer":
        self.trigram(1).young()
        self.trigram(0).young()
        return self

    def update_inner_values(self) -> None:
        self.trigram(1).value = self.value // 8
        self.trigram(0).value = self.value % 8
        self.trigram(1).update_inner_values()
        self.trigram(0).update_inner_values()

    def update_outer_values(self) -> None:
        self.value = self.trigram(0).value + self.trigram(1).value * 8

    @classmethod
    def set_current_sequence(cls, sequence: int) -> None:
        cls._nCurrentSequence = sequence

    @classmethod
    def set_current_ratio(cls, ratio: int) -> None:
        cls._nCurrentRatio = ratio

    @classmethod
    def set_current_label(cls, label: int) -> None:
        cls._nCurrentLabel = label

    def get_label(self) -> str:
        return Sequences.strHexagramLabels[self.get_current_label()][self.value]

    def get_moving(self) -> bool:
        return self.trigram(0).is_moving or self.trigram(1).is_moving

    def get_current_sequence(self) -> int:
        return CHexagramValueSequencer._nCurrentSequence

    def get_current_ratio(self) -> int:
        return CHexagramValueSequencer._nCurrentRatio

    def get_current_label(self) -> int:
        return CHexagramValueSequencer._nCurrentLabel

    def hexagram_id(self, use_value: bool = False) -> str:
        base = f"{(self.value if use_value else self.sequence + 1):2d}"
        if not self.is_moving:
            return base
        suffix = ''.join(
            str(line_index + 1)
            for line_index in range(6)
            if self.trigram(line_index // 3).line(line_index % 3).is_moving
        )
        return f"{base}.{suffix}"

    def describe_primary(self, use_value: bool = False) -> str:
        value_str = f" ({self.value_str})" if use_value else ""
        return f"{self.hexagram_id()} {self.label}{value_str}"

    def describe_secondary(self, use_value: bool = False) -> str:
        if not self.is_moving:
            return ""
        secondary = CHexagramValueSequencer(copy_from=self)
        secondary.move()
        value_str = f" ({secondary.value_str})" if use_value else ""
        return f"{secondary.hexagram_id()} {secondary.label}{value_str}"

    def describe_cast(self, use_value: bool = False) -> str:
        primary = self.describe_primary(use_value)
        secondary = self.describe_secondary(use_value)
        return f"{primary} > {secondary}" if secondary else primary

    def _save_moving(self) -> List[List[bool]]:
        moving = [[False, False, False], [False, False, False]]
        for t in range(2):
            for l in range(3):
                line_value = self.trigram(t).line(l).value
                if line_value in {0, 3}:
                    moving[t][l] = True
        return moving

    def _restore_moving(
        self,
        moving: List[List[bool]],
        inverse_trigram: bool,
        inverse_line: bool,
    ) -> None:
        for t in range(2):
            for l in range(3):
                if not moving[t][l]:
                    continue
                line_index = l
                if inverse_line:
                    if line_index == 0:
                        line_index = 2
                    elif line_index == 2:
                        line_index = 0
                trigram_index = 1 - t if inverse_trigram else t
                self.trigram(trigram_index).line(line_index).old()
